### ISSUE TITLE
Openmm without cudatoolkit dependency?

### ISSUE THREAD

**Issue by jennyfothergill** [openmm/openmm#3059]
(Forgive me if I am posting this in the wrong forum.) I am using openmm in a container built on [continuumio/miniconda3:4.9.2](https://hub.docker.com/layers/continuumio/miniconda3/4.9.2/images/sha256-7838d0ce65783b0d944c19d193e2e6232196bada9e5f3762dc7a9f07dc271179?context=explore). To my knowledge, I don't have cuda libraries in this container, yet openmm is pulling in cudatoolkit. It's not a blocking issue, but cudatoolkit comprises over half of the space of the container.
Is there a way to request openmm without cuda support?

**Comment by j-wags** [comment]
Hi @jennyfothergill. The Linux conda packages for OpenMM 7.5.0 on `conda-forge` have `cudatoolkit` as a hard requirement.

If you want to avoid the cudatoolkit requirement, I know of two alternate routes:
1) Use openmm 7.4.2 (`conda install -c conda-forge -c omnia openmm=7.4.2`)
2) Use openmm nightly builds (`conda install -c conda-forge -c omnia-dev openmm`)

**Comment by jchodera** [comment]
Can you force-uninstall `cudatoolkit` after installing the `conda-forge` openmm (but within the same docker step, so it doesn't increase the image size)?

**Comment by jennyfothergill** [comment]
@j-wags thank you! I will give that a try.
@jchodera unfortunately, uninstalling cudatoolkit will uninstall openmm

**Comment by jchodera** [comment]
I believe you can use the `--force` argument:
![image](https://user-images.githubusercontent.com/3656088/110841221-7e25b500-825a-11eb-8409-64e3fbe43d2f.png)

(Source: [stackoverflow](https://stackoverflow.com/questions/38459186/conda-uninstall-one-package-and-one-package-only))

**Comment by jennyfothergill** [comment]
@jchodera ah, OK, that may be a good solution!

**Comment by jchodera** [comment]
@jaimergp would know more if there are other ways to install without requiring `cudatoolkit`

**Comment by peastman** [comment]
> Use openmm 7.4.2 (conda install -c conda-forge -c omnia openmm=7.4.2)
> Use openmm nightly builds (conda install -c conda-forge -c omnia-dev openmm)

I wouldn't recommend either of those.  7.4.2 is out of date, and nightly builds should never be used for production work.

We should consider creating a "no CUDA" package that doesn't list it as a dependency.  I've run into a lot of issues lately, where conda insists on installing a different toolkit than the one I want to use.  Having conda install the toolkit was supposed to prevent problems, but it creates lots of new problems in the process.

**Comment by jennyfothergill** [comment]
OK, force removing cudatoolkit does work, and I am still able to import openmm. (I can't confirm if everything is functioning correctly yet)

but yes, if a no CUDA build is released, I would be interested! (and my container sizes would thank you 😄) I'm going to bring in @bdice because he recently set up smart hoomd builds on conda. He may have some insight into this.

**Comment by jchodera** [comment]
> We should consider creating a "no CUDA" package that doesn't list it as a dependency. I've run into a lot of issues lately, where conda insists on installing a different toolkit than the one I want to use. Having conda install the toolkit was supposed to prevent problems, but it creates lots of new problems in the process.

Should we open an issue here?
https://github.com/conda-forge/openmm-feedstock

**Comment by bdice** [comment]
For context, I made [this PR to conda-forge](https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/144) to make it possible to do autoselection of CPU/GPU builds using the `__cuda` metapackage that is present for new-ish versions of conda (something like 4.8+, off the top of my head). The presence of the `__cuda` metapackage indicates that a GPU with CUDA support is available on the current system. By pinning `__cuda` as a runtime requirement for the GPU builds, CPU-only systems will not be able to match GPU-enabled builds and thus install CPU-only builds (which can then drop the `cudatoolkit` requirement). 

Example from HOOMD:
https://github.com/conda-forge/hoomd-feedstock/blob/3143311fcb5071de725509d04158e4b38c6fc403/recipe/meta.yaml#L48-L49

If a CPU-only build (`cuda_compiler_version == "None"`) is added to the openmm feedstock, then it should be straightforward to enable the automatic detection of whether to use CPU/GPU builds based on the hardware in the system.

**Comment by jaimergp** [comment]
Problem is that we don't want to distribute CPU only builds. In other words, CUDA support must be present. We would just omit the `cudatoolkit` (heavy) dependency.

The only solutions I see:

a. Providing a `cudatoolkit-info` metapackage that only lists which CUDA version was used to compile the package. OpenMM variants would be available to target `cudatoolkit` and/or `cudatoolkit-info`. This might add extra complications in the CF ecosystem I am not anticipating.
b. Creating OpenMM metapackages: `openmm-base` and `openmm`. `openmm-base` would not list `cudatoolkit` as a dependency. `openmm` would depend on `openmm-base` and the corresponding `cudatoolkit` (similar to what `matplotlib` does with `qt`). CUDA version metadata would need to be provided either in the build string or through a metapackage.

A and B are somehow non exclusive, but this is a conversation that concerns the whole CF ecosystem, and we should probably add this to the `cudatoolkit` feedstock.

**Comment by mattwthompson** [comment]
Option b) sounds like an excellent solution for this problem. What would the broader implications for other packages be?

**Comment by jaimergp** [comment]
Possibly the existence of that metapackage. We'd just need to reach an agreement with the conda-forge core team about whether this is a good idea and, if so, how to implement this. The CUDA ecosystem has a few moving pieces already, so adding more can get tricky.

**Comment by bdice** [comment]
@jaimergp Can you give me more information on the reasoning behind this comment? I’m unaware of the context.

> Problem is that we don't want to distribute CPU only builds. In other words, CUDA support must be present.

If it’s possible, splitting into CPU/GPU builds like HOOMD could be a viable alternative that doesn’t require a metapackage. The approach in arrow-cpp-proc would be another option that doesn’t introduce a “base” package but does have a metapackage “arrow-cpp-proc” for selecting the build type.

**Comment by jaimergp** [comment]
Back when we started to migrate OpenMM to conda-forge, it was decided that the resulting package should always be distributed with all the platform plugins. @peastman might have some more comments on this.

**Comment by peastman** [comment]
OpenMM is rarely run on computers without GPUs.  There are two main reasons someone might want to avoid downloading the CUDA toolkit. 1) They already have a toolkit installed on their computer.  2) They have a non-NVIDIA GPU, so they're going to use OpenCL.  In either of those cases, they still need to get a build that includes GPU support.  A CPU-only build would have little value, and it would create a large risk of people accidentally ending up with a build that can't use their GPU.

**Comment by j-wags** [comment]
For what it's worth, OpenFF uses OpenMM for most of its energy evaluations, which are currently mostly single point energies and minimizations of small systems. So we'd make heavy use of a small CPU-only build :-)

**Comment by jennyfothergill** [comment]
Our lab would also be very happy to have a CPU-only build. :)

**Comment by jchodera** [comment]
@peastman : I don't think we have any statistics to support the idea that OpenMM is rarely run on computers without GPUs because we don't do any kind of monitoring of this. Given its embedding with QCFractal and the Open Force Field Toolkit, and the number of other applications that use it, it is quite possible that it's the other way around.

**Comment by mattwthompson** [comment]
There are some libraries that use parts of OpenMM's API that don't benefit from GPUs, like how Foyer's `Forcefield` class [subclasses](https://github.com/mosdef-hub/foyer/blob/e80015c98324b3eab77457fb35df98c2dcd39cd8/foyer/forcefield.py#L369) out from OpenMM's. In cases like these, the CUDA toolkit is dead weight. What are these users and developers recommended to do? From what I can tell, most of the source code is MIT licensed and allows (with some terms) for bits to be copied out and repackaged elsewhere - are there other solutions that are simpler to implement?

**Comment by swails** [comment]
We also use OpenMM for QM/MM stuff and have actually created a CPU-only conda package for internal use.

That said, in an ideal world the GPU platforms would "just work" if a compatible toolkit was available, but the CPU platform is more than fast enough to make the MM part negligible in speed.

**Comment by proteneer** [comment]
+1 to CPU only build. We use it here too.

**Comment by peastman** [comment]
> In cases like these, the CUDA toolkit is dead weight.

Agreed, that's why it would be useful to have a version that doesn't install a toolkit.  All the GPU related libraries in OpenMM itself add up to only a few MB total.  There's no significant benefit to leaving them out, and a very significant cost (the risk of someone getting the wrong version).  Being able to skip the 1 GB toolkit download is a much bigger benefit.

**Comment by gdonval** [comment]
I believe this would not be as much of a problem if cuda wasn't the size of a full OS:

``` 
cudatoolkit-11.2.2   | 877.3 MB  | ##                                    |   6%
```

I believe all AMD card users, people looking for single-point calculations and their kittens are extremely sad.

More seriously, providing seperate OpenCL and a CUDA builds would probably make sense. If the code allows for it, dynamically loading CUDA if available would solve the problem too. People could force not to install cudatoolkit by doing something along the line of:

```
conda create -n openmm -c conda-forge -c omnia openmm nocuda
```

This would be a virtual package exactly in the same fashion as `nomkl` exists today to for the use of an alternative blas implementation (generally OpenBLAS). This way, the default would install everything (and that's OK) but there is a way to opt out.

**Comment by swails** [comment]
> If the code allows for it, dynamically loading CUDA if available would solve the problem too.

This already happens.  All plugins, platforms, etc. are dynamically loaded and made available if they "work."  If you deploy it to a machine without a working CUDA installation, you simply won't get the CUDA platform (but the rest will work).

The problem here is specifically (and exclusively) about the conda packaging in which CUDA is set as a hard dependency and gets pulled in automatically.  If you install OpenMM through conda without dependencies, you'd get a version that worked with the CPU and Reference platforms (at least).

The tension here is how do we provide a build that newbies can conda-install and get a GPU-performant version but _also_ satisfy more experienced users that want to drop the CUDA requirement.

**Comment by jaimergp** [comment]
Yesterday in the conda-forge community call we were talking about a different problem (RHEL's openssl problem, if you are curious), and one of the mentioned solutions was adding a higher priority channel label that contains empty packages (just the dep metadata) to let the system library take over.

This might be applicable here as well and would like to investigate it. I'll add it to my list and try to champion it in the coming weeks!

**Comment by peastman** [comment]
Thanks, that would be great!  So basically you would add the option `-c conda-forge/label/nocuda`, and it would install a "cudatoolkit" package that was actually just empty?

A concern with this is that the next time you did anything with conda and didn't include that label, it would decide to automatically switch that package over to the standard version.  It seems to be pretty free about switching packages between channels whenever it feels like it.

**Comment by jaimergp** [comment]
For permanent effects, you could change the default channels list in your `~/.condarc` so the `conda-forge/label/nocuda` channel sits above the normal `conda-forge`.

**Comment by gdonval** [comment]
> Thanks, that would be great! So basically you would add the option -c conda-forge/label/nocuda, and it would install a "cudatoolkit" package that was actually just empty?

As I alluded to, packages `A` and package `B` can both provide package `C` that can be used for dependency handling. That is what is happening with `nomkl` in conda for instance: it prevents `MKL` from being used (and use the next thing it can find). 

In our case, both `cudatoolkit` and `nocuda` could be a `optional-cuda` provider. `nocuda` is empty **but** `cudatoolkit` is preferred (i.e. used by default, always, unless another provider is explicitly installed). Package that always require `CUDA` can use `cudatoolkit` instead of `optional-cuda`.

That method avoids meddling with channels so I can easily and "naturally" have OpenMM with cudatoolkit in one conda env and OpenMM without in another environment (for whatever reason). The only difference is 

    conda create -n openmm-env1 -c conda-forge openmm

vs.

    conda create -n openmm-env2 -c conda-forge openmm nocuda
    
No further support is required. Just that fake package and cudatoolkit sharing the same virtual package.

**Comment by jaimergp** [comment]
For the record I opened an issue for this at https://github.com/conda-forge/cudatoolkit-feedstock/issues/61 but it seems that we are going to prioritise the upcoming split packages, which would alleviate this issue too.

The only thing I can do now, outside of conda-forge, is to provide an opt-in channel with empty cudatoolkits that mimic the exact same metadata. With strict priority they will shadow CF's cudatoolkit.

Use it like this:

```
conda create -n openmm-nocuda -c jaimergp/label/unsupported-cudatoolkit-shim -c conda-forge --strict-channel-priority openmm
```

As the label says, there's no support or functional expectations around this approach. I tried it locally and the full `cudatoolkit` was not downloaded.

**Comment by peastman** [comment]
Thanks!  It's good to have that as an option.

**Comment by e-mayo** [comment]
It seems that it only works for the Linux platform. I use OpenMM in a GPU-enabled cluster but sometimes I need to run short simulation on my laptop.

I'd like to know if there is a way to install OpenMM for windows without cudatoolkit. I even have a NVIDIA GPU, and I don't know why downloading the cudatoolkit (900MB) is mandatory.

**Comment by yuxuanzhuang** [comment]
Hi @jaimergp! With the release of OpenMM 8, the minimum supported version of cudatoolkit (of some OpenMM versions) has been updated to 11.8. I'm curious to know if it's feasible to update the shim channel to newer versions: https://conda.anaconda.org/jaimergp/label/unsupported-cudatoolkit-shim/linux-64. 

```bash
 openmm 8.1.1 py310h43b6314_1
 ----------------------------
 file name   : openmm-8.1.1-py310h43b6314_1.conda
 name        : openmm
 version     : 8.1.1
 build       : py310h43b6314_1
 build number: 1
 license     : LGPL-3.0-or-later
 subdir      : linux-64
 url         : https://conda.anaconda.org/conda-forge/linux-64/openmm-8.1.1-py310h43b6314_1.conda
 md5         : 9610a3cdd30a01759c60567132fea487
 timestamp   : 2024-02-02 17:14:29 UTC
 dependencies:
   - __glibc >=2.17
   - __glibc >=2.17,<3.0.a0
   - cudatoolkit >=11.8,<12
   - libgcc-ng >=12
   - libstdcxx-ng >=12
   - numpy >=1.22.4,<2.0a0
   - ocl-icd >=2.3.1,<3.0a0
   - ocl-icd-system
   - python >=3.10,<3.11.0a0
   - python_abi 3.10.* *_cp310
   - scipy
```

Alternatively, is there a better way to build OpenMM without cudatoolkit now?

**Comment by peastman** [comment]
If you use CUDA 12 or later, the install is much smaller.  Instead of a single cudatoolkit package there are now separate packages for each of the libraries it contains.  It only needs to install the few libraries OpenMM uses, not the whole toolkit.

**Comment by yuxuanzhuang** [comment]
> If you use CUDA 12 or later, the install is much smaller. Instead of a single cudatoolkit package there are now separate packages for each of the libraries it contains. It only needs to install the few libraries OpenMM uses, not the whole toolkit.

Thanks! For those who are interested, to install the lightweight version of OpenMM, one needs to specify cuda-version>=12.0. Otherwise, with some Python versions, conda might select the one that requires the cudatoolkit dependency.

```bash
conda install -c conda-forge cuda-version>=12.0 openmm
```
